fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/Binder.java

```diff
@@ -162,6 +153,76 @@ private GroupExpression extractGroupExpression(Pattern pattern, Group group) {
         }
     }
 
+    /**
+     * Expression binding for MULTI_JOIN, which contains only JOIN/SCAN nodes
+     * <p>
+     * NOTE: Why not match using regular pattern ?
+     * 1. Regular matching cannot bind the tree but only the partial expression. But MULTI_JOIN needs to extract
+     * the entire tree
+     * 2. Regular matching is extremely slow in this case since it needs to recursive descent the tree, build the
+     * binding state and check the expression at the same time. But MULTI_JOIN could enumerate the GE without any check
+     */
+    class MultiJoinBinder {
+
+        public OptExpression match(GroupExpression ge) {
+            // 1. Check if the entire tree is MULTI_JOIN
+            // 2. Enumerate GE
+            if (ge == null || !isMultiJoin(ge)) {
+                return null;
+            }
+
+            return enumerate(ge);
+        }
+
+        private OptExpression enumerate(GroupExpression ge) {
+            if (ge == null) {
+                return null;
+            }
+            List<OptExpression> resultInputs = Lists.newArrayList();
+
+            int groupExpressionIndex = 0;
+            while (groupExpressionIndex < ge.getInputs().size()) {
+                trace();
+                Group group = ge.getInputs().get(groupExpressionIndex);
+                OptExpression opt = enumerate(extractGroupExpression(group));
+
+                if (opt == null) {
+                    return null;
+                } else {
+                    resultInputs.add(opt);
+                }
+
+                groupExpressionIndex++;
+            }
+
+            return new OptExpression(ge, resultInputs);
+        }
+
+        private GroupExpression extractGroupExpression(Group group) {
+            int valueIndex = groupExpressionIndex.get(groupTraceKey);
```

Comment:
How this process prove enumerate all optExpression. like 
(GroupA(a1, a2, a3) join GroupB(b1, b2, b3)) join GroupC(c1, c2, c3) when the outer binder invoke `next()` method how ensure the `match()` method return a new OptExpression using this `enumerate()` method.

Reply:
The enumerate procedure follows the DFS way, search from root to leaf, and go back to the upper level if explored every expression in the group.